#Data Cleaning: credit_risk_dataset.csv
This section covers data loading and initial inspection.

Data Cleaning Steps:
1. Load the data: Import data into the notebook (using pandas)
2. Inspect the data: use commands like .head(), .info(), .describe() to understand column types, and spot obvious issues
3. Check for missing values: identify which columns have missing (null) entries using .isnull() .sum()
4. Handle missing values: decide how to address missing data - fill with mean/median/mode, use interpolation, or drop rows/columns as appropriate 
    a. if the data is normally distributed, we impute the missing values using the overall mean
    b. if the data is skewed to the left or the right, we impute the missing values using the overall median
    c. if the data is categorical, then we impute using the mode of that feature itself
5. Remove Duplicates: any duplicated rows that could skew analysis
6. Check and Correct Data Types: ensure that columns are in correct format (numbers are int/float, categorical/text are object)
7. Clean categorical variables: look for capitalizations, typos, inconsistent labels, standardize values as needed
8. Export Cleaned Data to CSV

In [ ]:
import os
print(os.getcwd())


/Users/paxiong/MilwaukeeFUSE25/data-cleaning


STEP 1: Load the data: Import data into the notebook (using pandas)

In [ ]:
#Import data into the notebook
import pandas as pd
credit_df = pd.read_csv('../data/credit_risk_dataset.csv')


I imported the csv file into a pandas dataframe in order to bring the raw data into Python so I can inspect, analyze, and process it easily.

STEP 2: Inspect the data: use commands like .head(), .info(), .describe() to understand column types, and spot obvious issues

In [ ]:
#Inspects the first few rows of the DataFrame#
credit_df.head()

person_age  person_income person_home_ownership  person_emp_length  \
0          22          59000                  RENT              123.0   
1          21           9600                   OWN                5.0   
2          25           9600              MORTGAGE                1.0   
3          23          65500                  RENT                4.0   
4          24          54400                  RENT                8.0   

  loan_intent loan_grade  loan_amnt  loan_int_rate  loan_status  \
0    PERSONAL          D      35000          16.02            1   
1   EDUCATION          B       1000          11.14            0   
2     MEDICAL          C       5500          12.87            1   
3     MEDICAL          C      35000          15.23            1   
4     MEDICAL          C      35000          14.27            1   

   loan_percent_income cb_person_default_on_file  cb_person_cred_hist_length  
0                 0.59                         Y                           3  
1                 0.10                         N                           2  
2                 0.57                         N                           3  
3                 0.53                         N                           2  
4                 0.55                         Y                           4

I used .head() to view the first few rows.  This quick snapshot helps me spot any obvious issues (strange headers, missing data, weird formatting).

STEP 3: Check for missing values: identify which columns have missing (null) entries using .isnull() .sum() 

In [9]:
# Check For Missing Values and Data Types 
# How many rows, and columns you have, data types for each column, how many non-null values are in each column#
credit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  object 
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  object 
 5   loan_grade                  32581 non-null  object 
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  object 
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.0+ MB


Only Columns That Need to be Cleaned:
person_emp_length: float - 895 missing values
loan_int_rate: float - 3,116 missing values

int64: integer numbers (whole #, like 5, 20, 100)
float64: decimal numbers (fractional like 4.5, 99.99)
object: text/string data (like 'RENT', 'B', 'N') or mixed data types

In [ ]:
# Check the stats for 'person_emp_length' column
credit_df['person_emp_length'].describe()

count    31686.000000
mean         4.789686
std          4.142630
min          0.000000
25%          2.000000
50%          4.000000
75%          7.000000
max        123.000000
Name: person_emp_length, dtype: float64

Choose median for missing values because it is less affected by the outliers (max: 123)

I ran .info() and .describe() to check the column's data types and stats.  This helsp me understand the type of data each column represents (number, text, category), the size of the dataset, where things might be missing or inconsistent.

STEP 4: Handle missing values: check stats, decide how to address missing data - fill with mean/median/mode, use interpolation, or drop rows/columns as appropriate

Next: Fill Missing Values in person_emp_length.

fillna() tells pandas to “fill in the blanks” (NAs).
credit_df['person_emp_length'].median() finds the middle value in your data.
inplace=True updates your DataFrame directly.

In [11]:
credit_df['person_emp_length'].fillna(credit_df['person_emp_length'].median(), inplace=True)

/var/folders/m8/v8x4rymd34b8cryfyvt6mrcw0000gn/T/ipykernel_57940/318750665.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  credit_df['person_emp_length'].fillna(credit_df['person_emp_length'].median(), inplace=True)


In [13]:
credit_df['person_emp_length'] = credit_df['person_emp_length'].fillna(credit_df['person_emp_length'].median())


^ This syntax is "safer" by "assigning" the column (while still finding the median and filling the blanks) 

I filled the missing values in 'person_emp_length' column with the median because the data had outliers (max: 123 years of employment), so the median is less affected by extreme numbers and gives a more reliable "typical" value.

Repeat STEP 4 for 'loan_int_rate' Column

In [ ]:
#look at the statistics of the 'loan_int_rate' column
credit_df['loan_int_rate'].describe()

count    29465.000000
mean        11.011695
std          3.240459
min          5.420000
25%          7.900000
50%         10.990000
75%         13.470000
max         23.220000
Name: loan_int_rate, dtype: float64

In [16]:
# Count Missing Values in 'loan_int_rate'
credit_df['loan_int_rate'].isnull().sum()

np.int64(3116)

In [ ]:
# All the missing values at once:
credit_df.isnull().sum()

person_age                       0
person_income                    0
person_home_ownership            0
person_emp_length                0
loan_intent                      0
loan_grade                       0
loan_amnt                        0
loan_int_rate                 3116
loan_status                      0
loan_percent_income              0
cb_person_default_on_file        0
cb_person_cred_hist_length       0
dtype: int64

In [18]:
credit_df['loan_int_rate'].describe()

count    29465.000000
mean        11.011695
std          3.240459
min          5.420000
25%          7.900000
50%         10.990000
75%         13.470000
max         23.220000
Name: loan_int_rate, dtype: float64

Data is normally distributed (max: 23.22 is not extreme/not highly skewed/does not have strong outliers).
Use the mean if the distribution is roughly symmetric/normal and does not have strong outlier.
Use the median if the distribution is heavily skewed or has strong outlier.

In [19]:
#Fill Missing Values in 'loan_int_rate' with Mean
credit_df['loan_int_rate'] = credit_df['loan_int_rate'].fillna(credit_df['loan_int_rate'].mean())


In [20]:
credit_df.isnull().sum()

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

I filled the missing values with the column mean.  This column was not strongly skewed, and mean and median were similiar.  Using the mean keeps the overall average interest rate correct for later analysis.

STEP 5: Check for duplicates and remove them to prevent "double-counting" people or loans

In [21]:
credit_df.duplicated().sum()

np.int64(165)

In [22]:
# Remove Duplicates
credit_df.drop_duplicates(inplace=True)

In [23]:
# Double-check that duplicates are removed
credit_df.duplicated().sum()

np.int64(0)

I used .duplicated().sum() to count rows, then dropped them.  Duplicated rows can cause double counting, which ruins BI reports and metrics.  Removing them ensures each entry is unique and results are accurate.

STEP 6: Check and Correct Data Types: ensure that columns are in correct format (numbers are int/float, categorical/text are object)

In [24]:
# Check Data Types
credit_df.dtypes

person_age                      int64
person_income                   int64
person_home_ownership          object
person_emp_length             float64
loan_intent                    object
loan_grade                     object
loan_amnt                       int64
loan_int_rate                 float64
loan_status                     int64
loan_percent_income           float64
cb_person_default_on_file      object
cb_person_cred_hist_length      int64
dtype: object

In [25]:
categorical_cols = [
    'person_home_ownership',
    'loan_intent',
    'loan_grade',
    'cb_person_default_on_file',
]
for col in categorical_cols:
    credit_df[col] = credit_df[col].astype('category') 

I ran .dtypes to review each column's types and converted text-based columns with a small set of repeated labels (like home ownership or loan grade) to category type.  From "object" (text) to "category" to make analysis more efficient, faster, and to help tools like PowerBI/Tableau recognize these as discrete groupings for visualizations/business summaries.

STEP 7: Check for Odd Values (ex. typos)

In [26]:
for col in categorical_cols:
    print(col, credit_df[col].unique())

person_home_ownership ['RENT', 'OWN', 'MORTGAGE', 'OTHER']
Categories (4, object): ['MORTGAGE', 'OTHER', 'OWN', 'RENT']
loan_intent ['PERSONAL', 'EDUCATION', 'MEDICAL', 'VENTURE', 'HOMEIMPROVEMENT', 'DEBTCONSOLIDATION']
Categories (6, object): ['DEBTCONSOLIDATION', 'EDUCATION', 'HOMEIMPROVEMENT', 'MEDICAL', 'PERSONAL', 'VENTURE']
loan_grade ['D', 'B', 'C', 'A', 'E', 'F', 'G']
Categories (7, object): ['A', 'B', 'C', 'D', 'E', 'F', 'G']
cb_person_default_on_file ['Y', 'N']
Categories (2, object): ['N', 'Y']


I reviewed each categorical column using '.unique()' to amek sure there were no typos/formatting mistakes/unexpected values.  All categories are properly labeled and formatted consistently.

STEP 8: Export Cleaned Data

In [27]:
credit_df.to_csv('cleaned_credit_risk_dataset.csv', index=False)

I exported the cleaned credit risk dataset as 'cleaned_credit_risk_dataset.csv' so it can be loaded into Tableau for 
business intelligence visualizations and dashboard creation. This ensures my analysis is based on clean, trustworthy data.